lib/compilers/sail.ts

```diff
@@ -0,0 +1,177 @@
+
+    constructor(info: PreliminaryCompilerInfo, env: CompilationEnvironment) {
+        super(info, env);
+
+        this.outputFilebase = 'model';
+    }
+
+    override optionsForFilter(filters: ParseFiltersAndOutputOptions, outputFilename: any) {
+        // By default this adds C compiler options (-g etc).
+
+        // Add a fake option so we know to compile to binary in `runCompiler()`.
+        // This is how most other compilers work (with something like `-S`), and
+        // the `BaseCompiler` isn't really set up to allow anything else.
+        if (filters.binary) {
+            return ["--link-to-binary"];
+        }
+        return [];
+    }
+
+    override async runCompiler(
+        compiler: string,
+        options: string[],
+        inputFilename: string,
+        execOptions: ExecutionOptionsWithEnv,
+        filters?: ParseFiltersAndOutputOptions,
+    ): Promise<CompilationResult> {
+        if (!execOptions) {
+            execOptions = this.getDefaultExecOptions();
+        }
+
+        const tmpDir = path.dirname(inputFilename);
+
+        if (!execOptions.customCwd) {
+            execOptions.customCwd = tmpDir;
+        }
+
+        // We can't use `filters` to decide whether to compile to binary
+        // since it isn't always set (e.g. when this is called from
+        // buildExecutable()). Instead we use a fake command line option.
+        const binary = options.includes("--link-to-binary");
+        options = options.filter(option => option !== "--link-to-binary");
+
+        const fullResult: CompilationResult = {
+            code: 0,
+            timedOut: false,
+            stdout: [],
+            stderr: [],
+            buildsteps: [],
+            inputFilename,
+        };
+
+        const sailResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'Sail to C',
+            compiler,
+            [...options, '-o', this.outputFilebase],
+            execOptions,
+        );
+
+        if (sailResult.code !== 0 || !binary) {
+            return fullResult;
+        }
+
+        fullResult.executableFilename = inputFilename + '.exe';
+
+        // Query the sail compiler via `sail -dir` to find out where the
+        // C runtime files are (`sail.c` etc).
+        const sailDirResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'Get Sail dir',
+            compiler,
+            ['-dir'],
+            execOptions,
+        );
+
+        if (sailDirResult.code !== 0) {
+            return fullResult;
+        }
+
+        const sailDir = sailDirResult.stdout.map(line => line.text).join('\n').trim();
+
+        // Now compile the C file to an executable.
+        const outputFilenameC = this.getOutputFilename(tmpDir, this.outputFilebase);
+        const outputFilenameExe = outputFilenameC + '.exe';
+
+        const compileResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'C to binary',
+            'cc',
```

Comment:
what is `cc` here? Presumably you're hoping this runs a C compiler? That may not be the case more generally as our nodes that run CE typically don't have anything but the dependencies we put on them. I think we'll need a configuration option here and to set this configurably, perhaps defaulting to `gcc` on normal properties, else `/opt/compiler-explorer/gcc/14.2.0/bin/gcc` or similar on the amazon properties.

Like I said in the general comments, this is a weakness of CE's design: programs that generate intermediate compileable results really should have a UI that lets the user pick which compiler operates on their output.

Reply:
Yeah it's basically any C compiler. It's not very picky really. I'll add the config option.

Reply:
I added it as `cCompiler`. I did do `compiler.$[id}.cCompiler` but that seemed a bit overkill since it will probably always be the same, and it didn't seem to fall back to `cCompiler`. I assumed that's how the config system works since you can set `objdumper` or `compiler.${id}.objdumper` and they both seem to work? I dunno.

It has an option at least now.